Define a parallel queue for all file system operations

a-sully opened the pull request:
Fixes #74 by specifying a parallel queue for all file system operations, not just locking (see https://github.com/whatwg/fs/pull/9/files#r1109679577)

- [x] At least two implementers are interested (and none opposed):
   * Chromium
   * WebKit
- [ ] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * N/A (unless there's a way to deterministically test race conditions that I don't know about)
- [ ] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chromium: ...
   * Gecko: …
   * WebKit: …
- [ ] [MDN issue](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) is filed: …


<!--
    This comment and the below content is programmatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/fs/95.html" title="Last updated on Jun 13, 2023, 4:32 PM UTC (ba5a295)">Preview</a> | <a href="https://whatpr.org/fs/95/8360b75...ba5a295.html" title="Last updated on Jun 13, 2023, 4:32 PM UTC (ba5a295)">Diff</a>

a-sully commented:
cc @jesup @szewai

annevk commented on index.bs:
```diff
@@ -90,24 +90,28 @@ A <dfn export id=file>file entry</dfn> additionally consists of
+     1. Otherwise, if |lock| is "`taken-shared`":
+       1. Increase |count| by one.
+       1. Return true.
+  1. Return false.
```
This doesn't really work. These steps end up running in another conceptual thread. So the return values are lost.

You need to communicate back to the main thread through tasks.

You also need to actually start the parallel queue as otherwise it won't be executing its steps. See https://html.spec.whatwg.org/multipage/infrastructure.html#starting-a-new-parallel-queue.

(Note that the example there mistakenly doesn't queue tasks. That's a bug that still needs to be fixed.)

a-sully replied:
Ah, I see now. I made the mistake of basing this off of the example :)

I've sketched out what lock acquisition + release could look like on a couple methods. Please let me know if there's a better way you'd recommend doing this. Once it looks good, I'll copy this pattern to all the relevant methods

Some thoughts:
- Instead of passing the promise directly, should the "take a lock" algorithm be passed an "onFailure" algorithm (which could then reject the promise)? Right now we always reject if the lock acquisition fails, but presumably we could add a method that doesn't
- Is there a better way to do this? I found https://github.com/whatwg/infra/issues/181, but that doesn't seem to have gone anywhere
- Is it correct to have `onLockReleasedAlgorithm` be an algorithm, or should it be a task?

annevk replied:
1. Yeah. Or a single algorithm that invoke with an argument to let it know success/failure.
2. Not currently.
3. No that's correct. The task is the eventual wrapper for the algorithm and some other state.

a-sully replied:
Changed to pass only a single algorithm. Thanks for the suggestion!

annevk commented on index.bs:
```diff
@@ -90,37 +90,51 @@ A <dfn export id=file>file entry</dfn> additionally consists of
+       1. Increase |count| by one.
+       1. [=Queue a task=] to |replyTaskSource| to run |onLockTakenAlgorithm|
+          and return.
+  1. [=Reject=] |p| with a "{{NoModificationAllowedError}}" {{DOMException}}.
```
You have to queue a task for this as well. But maybe it makes more sense that you don't pass the promise in to this algorithm and you have a single callback algorithm that you then pass success/failure to or some such?

a-sully replied:
Makes sense! The original algorithm returned true or false, but it now runs the passed-in algorithm with "`success`" and "`failure`" explicitly, which seems more clear. Let me know if there's a better way to do this!

annevk commented on index.bs:
```diff
@@ -90,37 +90,51 @@ A <dfn export id=file>file entry</dfn> additionally consists of
+  1. If |value| is "`exclusive`":
+     1. If |lock| is "`open`":
+       1. Set lock to "`taken-exclusive`".
+       1. [=Queue a task=] to |replyTaskSource| to run |onLockTakenAlgorithm|
```
We want to be using "queue a global task". Passing a global across the "in parallel" boundary is generally accepted, assuming you don't start pulling state of it.

I also don't think we need _replyTaskSoruce_. We can just always use the storage task source.

a-sully replied:
Ack

annevk reviewed (commented):
This looks better, but not quite there yet. I'll be out next week, but happy to look at this again in February.

a-sully commented on index.bs:
```diff
@@ -1249,8 +1272,10 @@ The <dfn method for=FileSystemSyncAccessHandle>close()</dfn> method steps are:
 1. [=file entry/lock/release|Release the lock=] on
-   [=this=]'s [=FileSystemSyncAccessHandle/[[file]]=].
+   [=this=]'s [=FileSystemSyncAccessHandle/[[file]]=] with |onLockReleasedAlgorithm|.
+1. Wait for |onLockReleasedAlgorithm| to run.
```
Now that releasing a lock is async, presumably we can't just post a task to release the lock and return. Here's my attempt at saying "wait until the lock is released to return from `close()` - let me know if there's a better way to do this. Thanks!

annevk replied:
I think it would be slightly clearer for _onLockReleasedAlgorithm_ to flip a boolean.

And then instead of wait you want pause as per https://html.spec.whatwg.org/multipage/webappapis.html#pause because we're inside an actual task and not in parallel.

(And also please ensure to never introduce such a thing outside of workers!)

a-sully replied:
> I think it would be slightly clearer for _onLockReleasedAlgorithm_ to flip a boolean.

Done!

> And then instead of wait you want pause as per https://html.spec.whatwg.org/multipage/webappapis.html#pause because we're inside an actual task and not in parallel.
> 
> (And also please ensure to never introduce such a thing outside of workers!)

You betcha :)

annevk commented on index.bs:
```diff
@@ -90,37 +90,51 @@ A <dfn export id=file>file entry</dfn> additionally consists of
+  1. If |value| is "`exclusive`":
+     1. If |lock| is "`open`":
+       1. Set lock to "`taken-exclusive`".
+       1. [=Queue a global task=] on the [=storage task source=] to run
```
We want queue a storage task here and elsewhere. It seems like _file_'s relevant global object would be the correct global to use.

a-sully replied:
Aha! I wasn't aware we could pull file's relevant global object like that. I think this is right now. Sorry for the confusion!

annevk commented on index.bs:
```diff
@@ -90,37 +90,51 @@ A <dfn export id=file>file entry</dfn> additionally consists of
+       1. [=Queue a global task=] on the [=storage task source=] to run
+          |resultAlgorithm| with a value of "`success`" and return.
+     1. Otherwise, if |lock| is "`taken-shared`":
+       1. Increase |count| by one.
```
```suggestion
       1. Increase |count| by 1.
```

annevk commented on index.bs:
```diff
@@ -90,37 +90,51 @@ A <dfn export id=file>file entry</dfn> additionally consists of
+  1. Let |lock| be the |file|'s associated [=file entry/lock=].
+  1. Let |count| be the |file|'s [=file entry/shared lock count=].
+  1. If |lock| is "`taken-shared`":
+    1. Decrease |count| by one.
```
```suggestion
    1. Decrease |count| by 1.
```

annevk commented on index.bs:
```diff
@@ -353,13 +367,17 @@ The <dfn method for=FileSystemFileHandle>createWritable(|options|)</dfn> method
   1. If |access| is not "{{PermissionState/granted}}",
      [=reject=] |result| with a "{{NotAllowedError}}" {{DOMException}} and abort.
```
This needs a task.

a-sully replied:
Just to clarify, you're suggesting that the promise rejection should happen on some task source, not the `[=file system entry/request access=]` algorithm - correct?

This pattern (rejecting a promise from a step run in parallel) is repeated many times throughout this file. Would it be alright to fix these in a follow-up?

Also, just to confirm, is the promise rejection in the `lockResultAlgorithm` steps below fine as-is because it's already running on the storage task source? Or does that also need to queue a task?

Side note: while looking at this, I realized that `[=a new promise=]` is supposed to take a Realm. This appears to only be done correctly [in one method](https://fs.spec.whatwg.org/#dom-filesystemhandle-issameentry) for this spec. I can also fix this in a follow-up...

a-sully replied:
> This pattern (rejecting a promise from a step run in parallel) is repeated many times throughout this file. Would it be alright to fix these in a follow-up?

For an idea of what this could look like: https://github.com/whatwg/fs/pull/9/files#diff-5e793325cd2bfc452e268a4aa2f02b4024dd9584bd1db3c2595f61f1ecf7b985R273-R276

annevk replied:
Yeah, I guess a follow-up is okay.

And yeah, if you already have a task you're not in parallel and you can resolve promises as you wish. (Though sometimes you might still want to queue a task, it really depends on what kind of ordering you are after.)

annevk replied:
Actually, this used to be synchronous but now it runs in parallel, so I think we need to change this. If something was already in parallel and makes this mistake I think it's fine to leave it for a follow-up.

a-sully replied:
> And yeah, if you already have a task you're not in parallel and you can resolve promises as you wish. (Though sometimes you might still want to queue a task, it really depends on what kind of ordering you are after.)

Awesome, thanks for the clarification!

> Actually, this used to be synchronous but now it runs in parallel, so I think we need to change this. If something was already in parallel and makes this mistake I think it's fine to leave it for a follow-up.

SGTM. Let me know if the pattern I sketched out in https://github.com/whatwg/fs/pull/9#discussion_r1107406159 is appropriate here

a-sully replied:
This now runs in the `file system queue`

annevk reviewed (commented):
Mainly lots of nits, but this looks to be heading in the right direction to me!

annevk commented on index.bs:
```diff
@@ -90,37 +90,54 @@ A <dfn export id=file>file entry</dfn> additionally consists of
 
+A user agent has an associated <dfn>file system lock queue</dfn> which is the
+result of [=starting a new parallel queue=]. This queue is to be used for all
+[=tasks=] involving a [=file entry/lock=].
```
s/tasks/operations/? (The tasks are not the essential part after all, they're just for communicating results.)

a-sully replied:
Done

annevk commented on index.bs:
```diff
@@ -90,37 +90,54 @@ A <dfn export id=file>file entry</dfn> additionally consists of
-     1. Increase |count| by one.
-     1. Return true.
-1. Return false.
+To <dfn for="file entry/lock">take</dfn> a [=file entry/lock=] with a |value| of
```
Follow-up: maybe it would be clearer if we called this "take a lock". The current phrasing kinda gives the impression lock might be an argument.

a-sully replied:
Ack. I'll do this in a follow-up

annevk commented on index.bs:
```diff
@@ -90,37 +90,54 @@ A <dfn export id=file>file entry</dfn> additionally consists of
+       1. Set lock to "`taken-exclusive`".
+       1. [=Queue a storage task=] given |file|'s [=relevant global object=] and
+          |resultAlgorithm| with a value of "`success`".
+       1. Return.
```
We use "Abort these steps" in "in parallel" algorithms to not confuse readers they might actually return something useful to someone.

annevk commented on index.bs:
```diff
@@ -353,13 +370,17 @@ The <dfn method for=FileSystemFileHandle>createWritable(|options|)</dfn> method
+     runs these steps:
+    1. If |lockResult| is "`failure`", [=reject=] |result| with a
+       "{{NoModificationAllowedError}}" {{DOMException}} and abort.
+    1. Let |stream| be the result of [=create a new FileSystemWritableFileStream|creating a new FileSystemWritableFileStream=]
```
This really needs to happen from a task. Same below.

a-sully replied:
Also now happens on a task on the `file system queue`

annevk commented on index.bs:
```diff
@@ -1249,8 +1275,11 @@ The <dfn method for=FileSystemSyncAccessHandle>close()</dfn> method steps are:
 1. If [=this=]'s [=[[state]]=] is "`closed`", return.
 1. Set [=this=]'s [=[[state]]=] to "`closed`".
+1. Set |lockReleased| to false.
+1. Let |onLockReleasedAlgorithm| be this step: Set |lockReleased| to true.
```
```suggestion
1. Let |onLockReleasedAlgorithm| be this step: set |lockReleased| to true.
```

annevk commented on index.bs:
```diff
@@ -1249,8 +1275,11 @@ The <dfn method for=FileSystemSyncAccessHandle>close()</dfn> method steps are:
 1. [=file entry/lock/release|Release the lock=] on
-   [=this=]'s [=FileSystemSyncAccessHandle/[[file]]=].
+   [=this=]'s [=FileSystemSyncAccessHandle/[[file]]=] with |onLockReleasedAlgorithm|.
+1. [=Pause=] until |lockReleased| is set to true.
```
```suggestion
1. [=Pause=] until |lockReleased| is true.
```

annevk reviewed (commented):
With some more task queueing I think this can land.

a-sully commented:
Updated to post all file system operations to a "file system queue", as we discussed offline. Lock take/release is once again sync (and must be called from that queue). Please feel free to nit-pick this as much as you please, since this pattern will be applied to all the other methods. Thanks!

annevk commented on index.bs:
```diff
@@ -90,38 +90,47 @@ A <dfn export id=file>file entry</dfn> additionally consists of
    1. If |lock| is "`open`":
      1. Set lock to "`taken-exclusive`".
-     1. Return true.
+     1. Return a value of "`success`".
```
```suggestion
     1. Return "`success`".
```

annevk commented on index.bs:
```diff
@@ -90,38 +90,47 @@ A <dfn export id=file>file entry</dfn> additionally consists of
      1. Set |lock| to "`taken-shared`".
      1. Set |count| to 1.
-     1. Return true.
+     1. Return a value of "`success`".
```
```suggestion
     1. Return "`success`".
```

annevk commented on index.bs:
```diff
@@ -90,38 +90,47 @@ A <dfn export id=file>file entry</dfn> additionally consists of
-     1. Return true.
-1. Return false.
+     1. Increase |count| by 1.
+     1. Return a value of "`success`".
+1. Return a value of "`failure`".
```
```suggestion
     1. Return "`success`".
1. Return "`failure`".
```

annevk commented on index.bs:
```diff
@@ -90,38 +90,47 @@ A <dfn export id=file>file entry</dfn> additionally consists of
+     1. Return a value of "`success`".
+1. Return a value of "`failure`".
+
+Note: These steps must be run on the [=file system queue=].
```
You cannot require things in notes. And here we just need a statement of fact.
```suggestion
Note: These steps have to be run on the [=file system queue=].
```

annevk commented on index.bs:
```diff
@@ -90,38 +90,47 @@ A <dfn export id=file>file entry</dfn> additionally consists of
   1. If |count| is 0, set |lock| to "`open`".
 1. Otherwise, set |lock| to "`open`".
 
+Note: These steps must be run on the [=file system queue=].
```
```suggestion
Note: These steps have to be run on the [=file system queue=].
```

annevk reviewed (commented):
The overall setup here seems sound from a quick skim. The main problem is that "reject", "resolve", and "create a new X" are operations that need to happen on the main thread. And as such you need to do the "queue a storage task" dance for them.

Note that you don't necessarily need to replace each individually. You might be able to restructure things a bit so do all the in parallel work first and safe the results in variables and then queue a storage task that does the rejecting/resolving as per the variables.

Also, what kind of exception can "request access" throw, out of curiosity? Since we get that exception while in parallel, forwarding it probably doesn't entirely work. If we could normalize that to it not returning "granted" that would be better.

a-sully commented:
> The overall setup here seems sound from a quick skim. The main problem is that "reject", "resolve", and "create a new X" are operations that need to happen on the main thread. And as such you need to do the "queue a storage task" dance for them.

Oh I see... Question - does this work? If not, why not?
```
1. [=Queue a storage task=] with [=this=]'s [=relevant global object=] to
   [=enqueue the following steps=] to the [=file system queue=]:
   1. ...
```

> Note that you don't necessarily need to replace each individually. You might be able to restructure things a bit so do all the in parallel work first and safe the results in variables and then queue a storage task that does the rejecting/resolving as per the variables.
> 
> Also, what kind of exception can "request access" throw, out of curiosity? Since we get that exception while in parallel, forwarding it probably doesn't entirely work. If we could normalize that to it not returning "granted" that would be better.

Unfortunately I think this is needed, since it can reject with a `SecurityError`. See https://wicg.github.io/file-system-access/#permissions

annevk commented:
> Question - does this work? If not, why not?

Because you still end up in parallel when doing "resolve", etc. You need some things to happen "in parallel" and then after those things you queue a task for things to happen on the main thread. Such as resolving a promise or creating a JS object.

> Unfortunately I think this is needed, since it can reject with a SecurityError.

Hmm, I guess we should redesign this somewhat whereby "request access" returns some kind of internal value and we then later turn that into the appropriate exception.

a-sully commented:
> > Question - does this work? If not, why not?
> 
> Because you still end up in parallel when doing "resolve", etc. You need some things to happen "in parallel" and then after those things you queue a task for things to happen on the main thread. Such as resolving a promise or creating a JS object.

Ah yes, looking at this with fresh eyes I see the problem. I've updated a few methods - if that pattern looks good, I'll copy it everywhere else

> > Unfortunately I think this is needed, since it can reject with a SecurityError.
> 
> Hmm, I guess we should redesign this somewhat whereby "request access" returns some kind of internal value and we then later turn that into the appropriate exception.

This is a bit tricky because the [WICG spec considers file system access](https://wicg.github.io/file-system-access/#dom-permissionname-file-system) a "[powerful feature](https://w3c.github.io/permissions/#dfn-powerful-feature)", and powerful features have [permission request algorithms](https://wicg.github.io/permissions-request/#permission-request-algorithm) that are allowed to throw. I've added an explicit warning that this spec doesn't comply with that pattern, which seems like it should be good enough?

Anyways, I've updated the access checks to return either a PermissionState or an error code that can be turned into a DomException (though I wonder if _returning_ (as opposed to throwing) a DomException directly is an allowed pattern?)

If this all seems reasonable, I'll file an issue to update the WICG spec (and update all the instances in this spec which assume these algorithms can throw)

a-sully commented on index.bs:
```diff
@@ -61,16 +61,31 @@ different storage mechanism with a different API for such files. The entry point
+listed in the [=error names table=].
+Unless specified otherwise it returns "{{PermissionState/denied}}".
+
+<p class=warning> Dependent specifications may consider this API a
+[=powerful feature=]. However, unlike other [=powerful features=] whose
+[=permission request algorithm=] may throw, [=/file system entry=]'s
+[=file system entry/query access=] and [=file system entry/request access=]
+algorithms must run [=in parallel=] on the [=file system queue=] and are
+therefore not allowed to throw. Instead, the caller is expected to [=/reject=]
+as appropriate should these algorithms return an [=exception/error name=].
```
This really should go in WICG spec, but since we're doing something a bit funky it might be useful to have here, too?

annevk replied:
If they run in parallel the caller will also be in parallel. They wouldn't be able to immediately reject something.

But also I'm not a big fan of this design. Perhaps these should return a struct?

a-sully replied:
> If they run in parallel the caller will also be in parallel. They wouldn't be able to immediately reject something.

Updated to say `the caller is expected to [=queue a storage task=] to [=/reject=], as appropriate`

... though while going through the whole spec I realized that there is one situation - the _asynchronous iterator initialization steps_ - where there is no promise to reject. See the other comment

> But also I'm not a big fan of this design. Perhaps these should return a struct?

Done. AFAICT a struct item cannot be "optional" (only algorithm parameters?) so "error name" is just empty most of the time

a-sully commented on index.bs:
```diff
@@ -1032,29 +1079,43 @@ To
+           Note: It is expected that this atomically updates the contents of the
+           file on disk being written to.
+
+        1. [=file entry/lock/release|Release the lock=] on
+           |stream|'s [=FileSystemWritableFileStream/[[file]]=].
```
Looking at this again - since we're running in a "storage task" at this point (albeit one posted from the "file system queue"), do we need to explicitly queue this to the "file system queue"?

annevk replied:
Yes. Because these steps run on the main thread, not the "file system queue" thread.

a-sully commented:
> if that pattern looks good, I'll copy it everywhere else

friendly ping @annevk. This infra PR is holding up the PRs specifying `move()` and `remove()`

annevk commented on index.bs:
```diff
@@ -494,25 +518,35 @@ The <dfn method for=FileSystemFileHandle>createWritable(|options|)</dfn> method
-     "{{NotFoundError}}" {{DOMException}} and abort.
+  1. If |access| is not "{{PermissionState/granted}}":
+    1. Set |requestAccessError| to |access| if |access| is an
+       [=exception/error name=]; otherwise, "{{NotAllowedError}}".
```
```suggestion
       [=exception/error name=]; otherwise "{{NotAllowedError}}".
```

annevk reviewed (commented):
Thanks for your patience. This is looking rather good. Except the access checks, those could do with some more design work.

a-sully commented on index.bs:
```diff
@@ -654,13 +730,17 @@ support for example recursive iteration.
+    1. Set |accessErrorName| to |accessResult|'s
+       [=file system access result/error name=] if it is not
+       « the empty string »; otherwise "{{NotAllowedError}}".
+    1. [=Throw=] an |accessErrorName| {{DOMException}}.
```
We've specified that `query access` requires going in parallel on the `file system queue`, but if we get an exception there's no associated promise to reject...

... @mkruisselbrink do you know if this is this even necessary? Or is it enough to check `query access` in just the "get the next iteration result" algorithm below?

annevk replied:
It doesn't seem to do anything based on these steps. Per https://webidl.spec.whatwg.org/#idl-async-iterable we can omit it I think.

cc @domenic

a-sully replied:
Tentatively omitted it for now, pending input from others

a-sully commented on index.bs:
```diff
@@ -671,43 +751,54 @@ To [=get the next iteration result=] for a {{FileSystemDirectoryHandle}} |handle
+      1. Set |accessErrorName| to |accessResult|'s
+         [=file system access result/error name=] if it is not
+         « the empty string »; otherwise "{{NotAllowedError}}".
+      1. [=/Reject=] |promise| with an |accessErrorName| {{DOMException}}
+         and return |promise|.
+
+    1. If |directory| is `null`, [=/reject=] |result| with a
+       "{{NotFoundError}}" {{DOMException}} and abort these steps.
```
Note: this is the only "behavioral" change in this PR. Presumably the "NotAllowedError" should take precedence over the "NotAllowedError", just like everywhere else

a-sully commented on index.bs:
```diff
@@ -135,7 +176,7 @@ A <dfn export id=directory>directory entry</dfn> additionally consists of a [=/s
 Each member is either a [=/file entry=] or a [=/directory entry=].
 
-A [=/file system entry=] |entry| should be [=list/contained=] in the [=children=] of at most one
+A [=/file system entry=] |entry| should be [=list/contained=] in the [=directory entry/children=] of at most one
```
FYI: this fixes a spec build failure

a-sully reviewed (commented):
As promised, I've updated the whole spec to follow the pattern. Note that much of the diff is just formatting (e.g. adding an indentation)

I've manually double-checked that all promise rejections and resolutions now happen from the "storage task source" and all access checks are done in-parallel from the file system queue (though it's still possible I've missed something...)

annevk commented on index.bs:
```diff
@@ -61,16 +61,48 @@ different storage mechanism with a different API for such files. The entry point
+returns a [=/file system access result=].
+Unless specified otherwise it returns a [=/file system access result=] with a
+[=file system access result/permission state=] of "{{PermissionState/denied}}"
+and with a [=file system access result/error name=] of « the empty string ».
```
Wrapping this in « » is wrong as this is not a list. Applies quite a few times.

a-sully replied:
brain fart. This is fixed in the latest patch

annevk commented on index.bs:
```diff
@@ -61,16 +61,48 @@ different storage mechanism with a different API for such files. The entry point
+: <dfn for="file system access result">permission state</dfn>
+:: A {{PermissionState}}
+: <dfn for="file system access result">error name</dfn>
+:: A [=string=] which must be either « the empty string » or an
+   [=exception/error name=] listed in the [=error names table=].
+   If [=file system access result/permission state=] is
+   "{{PermissionState/granted}}" this must be « the empty string »
```
Can we make it so that when permission is denied it has to be an error name so we don't have to do defaulting below?

a-sully replied:
Done. This simplifies things quite a lot

Side note: is "iff" okay to use in specs? Happy to spell it out if not

a-sully commented:
@annevk friendly ping (this is holding up a number of other changes). If this looks good, feel free to merge

annevk commented on index.bs:
```diff
@@ -61,16 +61,51 @@ different storage mechanism with a different API for such files. The entry point
+returns a [=/file system access result=].
+Unless specified otherwise it returns a [=/file system access result=] with a
+[=file system access result/permission state=] of "{{PermissionState/denied}}"
+and with a [=file system access result/error name=] of the empty string.
```
```suggestion
and with an [=file system access result/error name=] of the empty string.
```

annevk commented on index.bs:
```diff
@@ -61,16 +61,51 @@ different storage mechanism with a different API for such files. The entry point
+returns a [=/file system access result=].
+Unless specified otherwise it returns a [=/file system access result=] with a
+[=file system access result/permission state=] of "{{PermissionState/denied}}"
+and with a [=file system access result/error name=] of the empty string.
```
```suggestion
and with an [=file system access result/error name=] of the empty string.
```

annevk reviewed (commented):
Found a couple remaining nits, generally looks good. Feel free to merge after addressing them. Though I should have time tomorrow if you'd rather I do it.

annevk reviewed (approved):
I meant to approve this to make that possible. Doh.

a-sully commented on index.bs:
```diff
@@ -61,16 +61,51 @@ different storage mechanism with a different API for such files. The entry point
+: <dfn for="file system access result">error name</dfn>
+:: A [=string=] which must be either the empty string or an
+   [=exception/error name=] listed in the [=error names table=].
+   Iff [=file system access result/permission state=] is
```
Ack. Converted into a requirement (i.e. "will" -> "must" or "should")

a-sully commented:
Thanks for the review! I'll merge this now so we can start making progress on the various PRs that are blocked on it. If you have more nits, I'm happy to address them in a follow-up

Just an FYI that I also updated the permission checks to use the new `[=DOMException/name=]` and "names table" added in https://github.com/whatwg/webidl/pull/1211 (which broke the refs we were adding)
